{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Taller 3. Filtrado de marcadores moleculares tipo SNP\n",
    "Genética de rasgos complejos 2024-01 - Docente Johana Carolina Soto Sedano\n",
    "\n",
    "__Hans D. Escobar H.__\n",
    "\n",
    "Las instrucciones están dadas para hacerlo en Excel. Lo haré con Python para practicar, también agregaré los pasos extra que considere pertinentes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path(\"../Private_Data/GBS.txt\")\n",
    "raw_df = pd.read_csv(str(data_path), sep=\"\\t\", index_col=0, dtype=str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Explorando los datos\n",
    "\n",
    "- Revisando los campos y estructura de la tabla:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['alleles', 'chrom', 'pos', 'strand', 'assembly', 'center', 'protLSID', 'assayLSID', 'panelLSID', 'QCcode']\n",
      "['G1', 'G2', 'G3', 'G4', 'G5', 'G6', 'G7', 'G8', 'G9', 'G10']\n",
      "['G11', 'G12', 'G13', 'G14', 'G15', 'G16', 'G17', 'G18', 'G19', 'G20']\n",
      "['G21', 'G22', 'G23', 'G24', 'G25', 'G26', 'G27', 'G28', 'G29', 'G30']\n",
      "['G31', 'G32', 'G33', 'G34', 'G35', 'G36', 'G37', 'G38', 'G39', 'G40']\n",
      "['G41', 'G42', 'G43', 'G44', 'G45', 'G46', 'G47', 'G48', 'G49', 'G50']\n",
      "['G51', 'G52', 'G53', 'G54', 'G55', 'G56', 'G57', 'G58', 'G59', 'G60']\n",
      "['G61', 'G62', 'G63', 'G64', 'G65', 'G66', 'G67', 'G68', 'G69', 'G70']\n",
      "['G71', 'G72', 'G73', 'G74', 'G75', 'G76', 'G77', 'G78', 'G79', 'G80']\n",
      "['G81', 'G82', 'G83', 'G84', 'G85', 'G86', 'G87', 'G88', 'G89', 'G90']\n",
      "['G91', 'G92', 'G93', 'G94', 'G95', 'G96', 'G97', 'G98', 'G99', 'G100']\n",
      "['G101', 'G102', 'G103', 'G104', 'G105', 'G106', 'G107', 'G108', 'G109', 'G110']\n",
      "['G111', 'G112', 'G113', 'G114', 'G115', 'G116', 'G117', 'G118', 'G119', 'G120']\n",
      "['G121', 'G122', 'G123', 'G124', 'G125', 'G126', 'G127', 'G128', 'G129', 'G130']\n",
      "['G131', 'G132', 'G133', 'G134', 'G135', 'G136', 'G137', 'G138', 'G139', 'G140']\n",
      "['G141', 'G142', 'G143', 'G144', 'G145', 'G146', 'G147', 'G148', 'G149', 'G150']\n"
     ]
    }
   ],
   "source": [
    "width = 10\n",
    "for line in range(0, len(raw_df.columns), width):\n",
    "    print(\"{}\".format(list(raw_df.columns[line : line + width])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de filas: 83830\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alleles</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>assembly</th>\n",
       "      <th>center</th>\n",
       "      <th>protLSID</th>\n",
       "      <th>assayLSID</th>\n",
       "      <th>panelLSID</th>\n",
       "      <th>QCcode</th>\n",
       "      <th>...</th>\n",
       "      <th>G141</th>\n",
       "      <th>G142</th>\n",
       "      <th>G143</th>\n",
       "      <th>G144</th>\n",
       "      <th>G145</th>\n",
       "      <th>G146</th>\n",
       "      <th>G147</th>\n",
       "      <th>G148</th>\n",
       "      <th>G149</th>\n",
       "      <th>G150</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SNP1</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2597267</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP2</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631717</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP3</th>\n",
       "      <td>C/A</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631718</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     alleles       chrom      pos strand assembly center protLSID assayLSID  \\\n",
       "rs                                                                            \n",
       "SNP1     C/T  ST4.03ch01  2597267      +      NaN    NaN      NaN       NaN   \n",
       "SNP2     A/G  ST4.03ch01  2631717      +      NaN    NaN      NaN       NaN   \n",
       "SNP3     C/A  ST4.03ch01  2631718      +      NaN    NaN      NaN       NaN   \n",
       "\n",
       "     panelLSID QCcode  ... G141 G142 G143 G144 G145 G146 G147 G148 G149 G150  \n",
       "rs                     ...                                                    \n",
       "SNP1       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   CC   CC   CC  \n",
       "SNP2       NaN    NaN  ...   AG   AA   AA   AA   AA   AA   AA   AA   AA   AA  \n",
       "SNP3       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   CC   CC   CC  \n",
       "\n",
       "[3 rows x 160 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Total de filas: {}\\n\".format(len(raw_df)))\n",
    "raw_df.head(n=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- ¿Hay loci multialélicos?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tipos de SNP presentes:\n",
      "\n",
      "['C/T' 'A/G' 'C/A' 'G/A' 'T/C' 'A/C' 'G/T' 'G/C' 'A/T' 'T/G' 'T/A' 'C/G']\n"
     ]
    }
   ],
   "source": [
    "print(\"Tipos de SNP presentes:\\n\\n{}\".format(raw_df[\"alleles\"].unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- ¿De que forma son los genotipos presentes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 genotipos únicos:\n",
      "\n",
      "{'AA', 'GT', 'AG', 'NN', 'TG', 'CA', 'CG', 'TC', 'CT', 'GA', 'TT', 'GC', 'AC', 'TA', 'GG', 'AT', 'CC'}\n"
     ]
    }
   ],
   "source": [
    "def get_unique_genotypes(df: pd.DataFrame):\n",
    "    genotypes = set()\n",
    "    for colname in df.columns[10:]:\n",
    "        genotypes.update(df[colname].unique())\n",
    "    return genotypes\n",
    "\n",
    "\n",
    "unique_genotypes = get_unique_genotypes(raw_df)\n",
    "print(\"{} genotipos únicos:\\n\\n{}\".format(len(unique_genotypes), unique_genotypes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Están los 16 esperados $card(\\{A, G, T, C\\} \\times \\{A, G, T, C\\})$, mas uno representando genotipo desconocido $(NN)$. No hay datos perdidos (NA)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Filtrando los datos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "class FilterDescription:\n",
    "    kept_markers: int\n",
    "    message: str\n",
    "\n",
    "    def __init__(self, message: str, kept_markers: int):\n",
    "        self.kept_markers = kept_markers\n",
    "        self.message = message\n",
    "\n",
    "    def __str__(self):\n",
    "        return f\"{self.message}: {self.kept_markers}\"\n",
    "\n",
    "\n",
    "filter_summary: list[FilterDescription] = list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Con base a la proporción de no genotipados"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alleles</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>assembly</th>\n",
       "      <th>center</th>\n",
       "      <th>protLSID</th>\n",
       "      <th>assayLSID</th>\n",
       "      <th>panelLSID</th>\n",
       "      <th>QCcode</th>\n",
       "      <th>...</th>\n",
       "      <th>G141</th>\n",
       "      <th>G142</th>\n",
       "      <th>G143</th>\n",
       "      <th>G144</th>\n",
       "      <th>G145</th>\n",
       "      <th>G146</th>\n",
       "      <th>G147</th>\n",
       "      <th>G148</th>\n",
       "      <th>G149</th>\n",
       "      <th>G150</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SNP1</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2597267</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP2</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631717</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP3</th>\n",
       "      <td>C/A</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631718</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP4</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631725</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP5</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631732</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83826</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100821</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83827</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100852</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83828</th>\n",
       "      <td>T/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100855</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TT</td>\n",
       "      <td>TA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83829</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61143072</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83830</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61143884</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83830 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         alleles       chrom       pos strand assembly center protLSID  \\\n",
       "rs                                                                       \n",
       "SNP1         C/T  ST4.03ch01   2597267      +      NaN    NaN      NaN   \n",
       "SNP2         A/G  ST4.03ch01   2631717      +      NaN    NaN      NaN   \n",
       "SNP3         C/A  ST4.03ch01   2631718      +      NaN    NaN      NaN   \n",
       "SNP4         G/A  ST4.03ch01   2631725      +      NaN    NaN      NaN   \n",
       "SNP5         C/T  ST4.03ch01   2631732      +      NaN    NaN      NaN   \n",
       "...          ...         ...       ...    ...      ...    ...      ...   \n",
       "SNP83826     A/G  ST4.03ch12  61100821      +      NaN    NaN      NaN   \n",
       "SNP83827     G/A  ST4.03ch12  61100852      +      NaN    NaN      NaN   \n",
       "SNP83828     T/A  ST4.03ch12  61100855      +      NaN    NaN      NaN   \n",
       "SNP83829     G/A  ST4.03ch12  61143072      +      NaN    NaN      NaN   \n",
       "SNP83830     C/T  ST4.03ch12  61143884      +      NaN    NaN      NaN   \n",
       "\n",
       "         assayLSID panelLSID QCcode  ... G141 G142 G143 G144 G145 G146 G147  \\\n",
       "rs                                   ...                                      \n",
       "SNP1           NaN       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   \n",
       "SNP2           NaN       NaN    NaN  ...   AG   AA   AA   AA   AA   AA   AA   \n",
       "SNP3           NaN       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   \n",
       "SNP4           NaN       NaN    NaN  ...   GA   GG   GG   GG   GG   GG   GG   \n",
       "SNP5           NaN       NaN    NaN  ...   CT   TT   TT   TT   TT   TT   TT   \n",
       "...            ...       ...    ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "SNP83826       NaN       NaN    NaN  ...   AA   AG   AA   AA   AA   AA   AA   \n",
       "SNP83827       NaN       NaN    NaN  ...   GG   GA   GG   GG   GG   GG   GG   \n",
       "SNP83828       NaN       NaN    NaN  ...   TT   TA   TT   TT   TT   TT   TT   \n",
       "SNP83829       NaN       NaN    NaN  ...   GA   GG   GA   GA   GG   GA   GG   \n",
       "SNP83830       NaN       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   \n",
       "\n",
       "         G148 G149 G150  \n",
       "rs                       \n",
       "SNP1       CC   CC   CC  \n",
       "SNP2       AA   AA   AA  \n",
       "SNP3       CC   CC   CC  \n",
       "SNP4       GG   GG   GG  \n",
       "SNP5       TT   CT   TT  \n",
       "...       ...  ...  ...  \n",
       "SNP83826   AA   AA   AA  \n",
       "SNP83827   GG   GG   GG  \n",
       "SNP83828   TT   TT   TT  \n",
       "SNP83829   GA   GG   GG  \n",
       "SNP83830   CC   CC   CC  \n",
       "\n",
       "[83830 rows x 160 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Preserva los que tienen igual o mayor proporción que threshold\n",
    "def missing_proportion_filter(df, threshold=0.85):\n",
    "    def filter_NN_and_NaN(row, threshold):\n",
    "        counts = row.iloc[10:].value_counts(normalize=True, dropna=False)\n",
    "        return counts.get(\"NN\", 0) + counts.get(pd.NA, 0) < threshold\n",
    "\n",
    "    return df[df.apply(filter_NN_and_NaN, axis=1, threshold=threshold)]\n",
    "\n",
    "\n",
    "missing_threshold = 0.85\n",
    "preprocessed_df = missing_proportion_filter(raw_df, missing_threshold)\n",
    "filter_summary.append(\n",
    "    FilterDescription(\n",
    "        f\"Marcadores con frecuencia de datos faltantes (NN o sin datos) < {1-missing_threshold:.3f}\",\n",
    "        len(preprocessed_df),\n",
    "    )\n",
    ")\n",
    "preprocessed_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Removiendo marcadores no informativos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Duplicados con segregación idéntica\n",
    "def identical_segregation_filter(df):\n",
    "    return df.drop_duplicates(subset=df.columns[10:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alleles</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>assembly</th>\n",
       "      <th>center</th>\n",
       "      <th>protLSID</th>\n",
       "      <th>assayLSID</th>\n",
       "      <th>panelLSID</th>\n",
       "      <th>QCcode</th>\n",
       "      <th>...</th>\n",
       "      <th>G141</th>\n",
       "      <th>G142</th>\n",
       "      <th>G143</th>\n",
       "      <th>G144</th>\n",
       "      <th>G145</th>\n",
       "      <th>G146</th>\n",
       "      <th>G147</th>\n",
       "      <th>G148</th>\n",
       "      <th>G149</th>\n",
       "      <th>G150</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SNP1</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2597267</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP2</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631717</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP3</th>\n",
       "      <td>C/A</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631718</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP4</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631725</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP5</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631732</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83826</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100821</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83827</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100852</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83828</th>\n",
       "      <td>T/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100855</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TT</td>\n",
       "      <td>TA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83829</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61143072</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83830</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61143884</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68189 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         alleles       chrom       pos strand assembly center protLSID  \\\n",
       "rs                                                                       \n",
       "SNP1         C/T  ST4.03ch01   2597267      +      NaN    NaN      NaN   \n",
       "SNP2         A/G  ST4.03ch01   2631717      +      NaN    NaN      NaN   \n",
       "SNP3         C/A  ST4.03ch01   2631718      +      NaN    NaN      NaN   \n",
       "SNP4         G/A  ST4.03ch01   2631725      +      NaN    NaN      NaN   \n",
       "SNP5         C/T  ST4.03ch01   2631732      +      NaN    NaN      NaN   \n",
       "...          ...         ...       ...    ...      ...    ...      ...   \n",
       "SNP83826     A/G  ST4.03ch12  61100821      +      NaN    NaN      NaN   \n",
       "SNP83827     G/A  ST4.03ch12  61100852      +      NaN    NaN      NaN   \n",
       "SNP83828     T/A  ST4.03ch12  61100855      +      NaN    NaN      NaN   \n",
       "SNP83829     G/A  ST4.03ch12  61143072      +      NaN    NaN      NaN   \n",
       "SNP83830     C/T  ST4.03ch12  61143884      +      NaN    NaN      NaN   \n",
       "\n",
       "         assayLSID panelLSID QCcode  ... G141 G142 G143 G144 G145 G146 G147  \\\n",
       "rs                                   ...                                      \n",
       "SNP1           NaN       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   \n",
       "SNP2           NaN       NaN    NaN  ...   AG   AA   AA   AA   AA   AA   AA   \n",
       "SNP3           NaN       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   \n",
       "SNP4           NaN       NaN    NaN  ...   GA   GG   GG   GG   GG   GG   GG   \n",
       "SNP5           NaN       NaN    NaN  ...   CT   TT   TT   TT   TT   TT   TT   \n",
       "...            ...       ...    ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "SNP83826       NaN       NaN    NaN  ...   AA   AG   AA   AA   AA   AA   AA   \n",
       "SNP83827       NaN       NaN    NaN  ...   GG   GA   GG   GG   GG   GG   GG   \n",
       "SNP83828       NaN       NaN    NaN  ...   TT   TA   TT   TT   TT   TT   TT   \n",
       "SNP83829       NaN       NaN    NaN  ...   GA   GG   GA   GA   GG   GA   GG   \n",
       "SNP83830       NaN       NaN    NaN  ...   CC   CC   CC   CC   CC   CC   CC   \n",
       "\n",
       "         G148 G149 G150  \n",
       "rs                       \n",
       "SNP1       CC   CC   CC  \n",
       "SNP2       AA   AA   AA  \n",
       "SNP3       CC   CC   CC  \n",
       "SNP4       GG   GG   GG  \n",
       "SNP5       TT   CT   TT  \n",
       "...       ...  ...  ...  \n",
       "SNP83826   AA   AA   AA  \n",
       "SNP83827   GG   GG   GG  \n",
       "SNP83828   TT   TT   TT  \n",
       "SNP83829   GA   GG   GG  \n",
       "SNP83830   CC   CC   CC  \n",
       "\n",
       "[68189 rows x 160 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preprocessed_df = identical_segregation_filter(preprocessed_df)\n",
    "filter_summary.append(\n",
    "    FilterDescription(\n",
    "        \"Marcadores únicos\",\n",
    "        len(preprocessed_df),\n",
    "    )\n",
    ")\n",
    "preprocessed_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Removiendo genotipos raros.\n",
    "\n",
    "No se cuenta con datos de calidad del genotipado, por lo que el MAF (Minor Allele Frequency) se calculara directamente de los genotipos llamados."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preserva los marcadores cuyo alelo menos frecuente esta en por lo menos 5% de la muestra\n",
    "# Estimation of Minor allele frequency from called genotypes\n",
    "def maf_from_called_filter(df: pd.DataFrame, threshold=0.05):\n",
    "    def get_maf_from_called(row: pd.Series, threshold):\n",
    "        allele_count = dict()\n",
    "        total = 0\n",
    "        for genotype, count in row[10:].value_counts().items():\n",
    "            if genotype == \"NN\":\n",
    "                continue\n",
    "            total += 2 * count\n",
    "            for allele in list(str(genotype)):\n",
    "                if allele in allele_count:\n",
    "                    allele_count[allele] += count\n",
    "                else:\n",
    "                    allele_count[allele] = count\n",
    "        return min(allele_count.values()) / total >= threshold\n",
    "\n",
    "    return df[df.apply(get_maf_from_called, axis=1, threshold=threshold)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alleles</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>assembly</th>\n",
       "      <th>center</th>\n",
       "      <th>protLSID</th>\n",
       "      <th>assayLSID</th>\n",
       "      <th>panelLSID</th>\n",
       "      <th>QCcode</th>\n",
       "      <th>...</th>\n",
       "      <th>G141</th>\n",
       "      <th>G142</th>\n",
       "      <th>G143</th>\n",
       "      <th>G144</th>\n",
       "      <th>G145</th>\n",
       "      <th>G146</th>\n",
       "      <th>G147</th>\n",
       "      <th>G148</th>\n",
       "      <th>G149</th>\n",
       "      <th>G150</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SNP2</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631717</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP4</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631725</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP5</th>\n",
       "      <td>C/T</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631732</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP6</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>2631734</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP7</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch01</td>\n",
       "      <td>3683816</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83825</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100816</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83826</th>\n",
       "      <td>A/G</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100821</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83827</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100852</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83828</th>\n",
       "      <td>T/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61100855</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TT</td>\n",
       "      <td>TA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP83829</th>\n",
       "      <td>G/A</td>\n",
       "      <td>ST4.03ch12</td>\n",
       "      <td>61143072</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>45851 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         alleles       chrom       pos strand assembly center protLSID  \\\n",
       "rs                                                                       \n",
       "SNP2         A/G  ST4.03ch01   2631717      +      NaN    NaN      NaN   \n",
       "SNP4         G/A  ST4.03ch01   2631725      +      NaN    NaN      NaN   \n",
       "SNP5         C/T  ST4.03ch01   2631732      +      NaN    NaN      NaN   \n",
       "SNP6         G/A  ST4.03ch01   2631734      +      NaN    NaN      NaN   \n",
       "SNP7         A/G  ST4.03ch01   3683816      +      NaN    NaN      NaN   \n",
       "...          ...         ...       ...    ...      ...    ...      ...   \n",
       "SNP83825     A/G  ST4.03ch12  61100816      +      NaN    NaN      NaN   \n",
       "SNP83826     A/G  ST4.03ch12  61100821      +      NaN    NaN      NaN   \n",
       "SNP83827     G/A  ST4.03ch12  61100852      +      NaN    NaN      NaN   \n",
       "SNP83828     T/A  ST4.03ch12  61100855      +      NaN    NaN      NaN   \n",
       "SNP83829     G/A  ST4.03ch12  61143072      +      NaN    NaN      NaN   \n",
       "\n",
       "         assayLSID panelLSID QCcode  ... G141 G142 G143 G144 G145 G146 G147  \\\n",
       "rs                                   ...                                      \n",
       "SNP2           NaN       NaN    NaN  ...   AG   AA   AA   AA   AA   AA   AA   \n",
       "SNP4           NaN       NaN    NaN  ...   GA   GG   GG   GG   GG   GG   GG   \n",
       "SNP5           NaN       NaN    NaN  ...   CT   TT   TT   TT   TT   TT   TT   \n",
       "SNP6           NaN       NaN    NaN  ...   GG   GG   GG   GG   GG   GG   GG   \n",
       "SNP7           NaN       NaN    NaN  ...   AA   AA   AA   AA   AA   AA   AA   \n",
       "...            ...       ...    ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "SNP83825       NaN       NaN    NaN  ...   AA   AG   AA   AA   AA   AA   AA   \n",
       "SNP83826       NaN       NaN    NaN  ...   AA   AG   AA   AA   AA   AA   AA   \n",
       "SNP83827       NaN       NaN    NaN  ...   GG   GA   GG   GG   GG   GG   GG   \n",
       "SNP83828       NaN       NaN    NaN  ...   TT   TA   TT   TT   TT   TT   TT   \n",
       "SNP83829       NaN       NaN    NaN  ...   GA   GG   GA   GA   GG   GA   GG   \n",
       "\n",
       "         G148 G149 G150  \n",
       "rs                       \n",
       "SNP2       AA   AA   AA  \n",
       "SNP4       GG   GG   GG  \n",
       "SNP5       TT   CT   TT  \n",
       "SNP6       GG   GA   GG  \n",
       "SNP7       AA   AA   AA  \n",
       "...       ...  ...  ...  \n",
       "SNP83825   AA   AA   AA  \n",
       "SNP83826   AA   AA   AA  \n",
       "SNP83827   GG   GG   GG  \n",
       "SNP83828   TT   TT   TT  \n",
       "SNP83829   GA   GG   GG  \n",
       "\n",
       "[45851 rows x 160 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maf_threshold = 0.05\n",
    "preprocessed_df = maf_from_called_filter(preprocessed_df, maf_threshold)\n",
    "filter_summary.append(\n",
    "    FilterDescription(\n",
    "        f\"Marcadores cuyo MAF >= {maf_threshold:.3f}\",\n",
    "        len(preprocessed_df),\n",
    "    )\n",
    ")\n",
    "preprocessed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Marcadores con frecuencia de datos faltantes (NN o sin datos) < 0.150: 83830\n",
      "Marcadores únicos: 68189\n",
      "Marcadores cuyo MAF >= 0.050: 45851\n"
     ]
    }
   ],
   "source": [
    "print(*filter_summary, sep=\"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
